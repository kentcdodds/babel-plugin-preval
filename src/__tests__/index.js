import path from 'path'
import pluginTester from 'babel-plugin-tester'
import plugin from '../'

const projectRoot = path.join(__dirname, '../../')

expect.addSnapshotSerializer({
  print(val) {
    return val.split(projectRoot).join('<PROJECT_ROOT>/')
  },
  test(val) {
    return typeof val === 'string'
  },
})

const error = code => ({code, error: true})
const noSnapshot = code => ({code, snapshot: false})
const fixture = filename => ({
  fixture: require.resolve(`./fixtures/${filename}`),
})

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {filename: __filename},
  tests: {
    'not preval': noSnapshot(
      'const x = notPreval`module.exports = "nothing"`;',
    ),
    'simple number': 'const x = preval`module.exports = 1`',
    'simple string': 'preval`module.exports = "foo"`',
    'simple function': `
      const y = preval\`
        module.exports = {
          booyah: function() {
            return "booyah"
          }
        }
      \`
    `,
    'a-bunch-of-checks': fixture('a-bunch-of-checks'),
    'dynamic error': error(
      `const x = preval\`module.exports = "\${dynamic}"\``,
    ),
    'import comment': 'import x from /* preval */ "./fixtures/compute-one.js"',
    'import comment (with extras)':
      'import x from /* preval */ /* this is extra stuff */ "./fixtures/compute-one.js"',
    'import comment (with extras before)':
      'import x from /* this is extra stuff */ /* preval */ "./fixtures/compute-one.js"',
    'invalid comment': noSnapshot(
      'import x from /* this is extra stuff */"./fixtures/compute-one.js";',
    ),
    'import string arg':
      'import x from /* preval("string argument") */ "./fixtures/identity.js"',
    'import object arg':
      'import x from /* preval({object: "argument", withFunction: () => {}}) */ "./fixtures/identity.js"',
    'import required arg':
      'import x from /* preval(require("./fixtures/compute-one")) */ "./fixtures/identity"',
    'simple require': 'const x = preval.require("./fixtures/compute-one")',
    'require with arge': 'const x = preval.require("./fixtures/identity", 3)',
    'require functions':
      'const x = preval.require("./fixtures/multiple-functions")',
    'absolute-path':
      'const x = preval.require("../__tests__/fixtures/nested/absolute-path")',
    'require with unknown arg': error(
      'const x = preval.require("./fixtures/identity", SOME_UNKNOWN_VARIABLE)',
    ),
    'require with arg but not a function': error(
      'const x = preval.require("./fixtures/compute-one", "should not be here...")',
    ),
    'simple comment': `
      // @preval
      module.exports = 1 + 2 - 1 - 1
    `,
    'with flow before': `
      // @flow
      // @preval
      module.exports = 1 + 2 - 1 - 1
    `,
    'with flow after': `
      // @preval
      // @flow
      module.exports = 1 + 2 - 1 - 1
    `,
    'export undefined': `
      // @preval
      let smth = {}
      module.exports = smth.UNDEFINED;
    `,
    'comment no contents': noSnapshot('// @preval'),
    'comment with only comment contents': noSnapshot(`
      // @preval
      /* comment */
    `),
    'handles transpiled modules (uses default export)': `
      let one = preval.require('./fixtures/es6-identity.js', 1)
    `,
  },
})
