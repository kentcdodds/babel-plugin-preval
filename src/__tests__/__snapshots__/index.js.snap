// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`2. preval 1`] = `

const x = preval\`module.exports = 1\`

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 1;

`;

exports[`3. preval 1`] = `

preval\`module.exports = "foo"\`

      ↓ ↓ ↓ ↓ ↓ ↓

"foo";

`;

exports[`4. preval 1`] = `

const y = preval\`
  module.exports = {
    booyah: function() {
      return "booyah"
    }
  }
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const y = { "booyah": function () {
    return "booyah";
  } };

`;

exports[`5. preval 1`] = `

const x = preval\`
  const fs = require('fs')
  const val = fs.readFileSync(require.resolve('./fixture1.md'), 'utf8')
  module.exports = {
    val,
    getSplit: function(splitDelimiter) {
      return x.val.split(splitDelimiter)
    }
  }
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const x = { "val": "# fixture\\n\\nThis is some file thing...\\n", "getSplit": function getSplit(splitDelimiter) {
    return x.val.split(splitDelimiter);
  } };

`;

exports[`6. preval 1`] = `

export default preval\`
  import path from 'path'
  export default path.join(__dirname, 'something/whatever.js').replace(__dirname, '')
\`

      ↓ ↓ ↓ ↓ ↓ ↓

export default { "default": "/something/whatever.js" };

`;

exports[`7. preval 1`] = `

const x = preval\`module.exports = "\${dynamic}"\`

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/index.js: Unable to determine the value of your preval string

`;

exports[`8. preval 1`] = `

import x from /* preval */ "./fixtures/compute-one.js"

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 1;

`;

exports[`9. preval 1`] = `

import x from /* preval */ "./fixtures/compute-one.js"

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 1;

`;

exports[`10. preval 1`] = `

import x from /* preval */ /* this is extra stuff */ "./fixtures/compute-one.js"

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 1;

`;

exports[`11. preval 1`] = `

import x from /* this is extra stuff */ /* preval */ "./fixtures/compute-one.js"

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 1;

`;

exports[`13. preval 1`] = `

import x from /* preval("string argument") */ "./fixtures/identity.js"

      ↓ ↓ ↓ ↓ ↓ ↓

const x = "string argument";

`;

exports[`14. preval 1`] = `

import x from /* preval({object: "argument", withFunction: () => {}}) */ "./fixtures/identity.js"

      ↓ ↓ ↓ ↓ ↓ ↓

const x = { "object": "argument", "withFunction": () => {} };

`;

exports[`15. preval 1`] = `

import x from /* preval(require("./fixtures/compute-one")) */ "./fixtures/identity"

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 1;

`;

exports[`16. preval 1`] = `

import x from /* preval(require("./fixtures/es6").default) */ "./fixtures/es6-identity"

      ↓ ↓ ↓ ↓ ↓ ↓

const x = "this string used exported default";

`;

exports[`17. preval 1`] = `

const x = preval.require("./fixtures/compute-one")

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 1;

`;

exports[`18. preval 1`] = `

const x = preval.require("./fixtures/identity", 3)

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 3;

`;

exports[`19. preval 1`] = `

const x = preval.require("./fixtures/multiple-functions")

      ↓ ↓ ↓ ↓ ↓ ↓

const x = { "a": function a() {
    return 'a';
  }, "b": function b() {
    return 'b';
  } };

`;

exports[`20. preval 1`] = `

const x = preval.require("../__tests__/fixtures/nested/absolute-path")

      ↓ ↓ ↓ ↓ ↓ ↓

const x = "<PROJECT_ROOT>/src/__tests__/fixtures/nested/absolute-path.js";

`;

exports[`21. preval 1`] = `

const x = preval.require("./fixtures/identity", SOME_UNKNOWN_VARIABLE)

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/index.js: preval cannot determine the value of an argument in preval.require

`;

exports[`22. preval 1`] = `

const x = preval.require("./fixtures/compute-one", "should not be here...")

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/index.js: \`preval.require\`-ed module (./fixtures/compute-one) cannot accept arguments because it does not export a function. You passed the arguments: should not be here...

`;

exports[`23. preval 1`] = `

// @preval
module.exports = 1 + 2 - 1 - 1

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
module.exports = 1;

`;

exports[`24. preval 1`] = `

// @preval
const ten = 9 + 1
module.exports = ten * 5

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
module.exports = 50;

`;

exports[`25. preval 1`] = `

// @flow
// @preval
module.exports = 1 + 2 - 1 - 1

      ↓ ↓ ↓ ↓ ↓ ↓

// @flow
// this file was prevaled
module.exports = 1;

`;

exports[`26. preval 1`] = `

// @preval
// @flow
module.exports = 1 + 2 - 1 - 1

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
// @flow
module.exports = 1;

`;

exports[`27. preval 1`] = `

// @preval
const name = 'Bob Hope'
const splitter = str => str.split(' ')
module.exports = splitter(name)

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
module.exports = ["Bob", "Hope"];

`;

exports[`28. preval 1`] = `

// @preval
const name = 'Bob Hope'
const splitter = str => str.split(' ')
const [first, last] = splitter(name)
module.exports = {first, last}

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
module.exports = { "first": "Bob", "last": "Hope" };

`;

exports[`29. preval 1`] = `

// @preval
module.exports = require("./fixtures/compute-one")

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
module.exports = 1;

`;

exports[`30. preval 1`] = `

// @preval
module.exports = require("./fixtures/identity")('hello world')

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
module.exports = "hello world";

`;

exports[`31. preval 1`] = `

// @preval
const id = require("./fixtures/identity")
const computeOne = require("./fixtures/compute-one")

const compose = (...fns) => fns.reduce((f, g) => a => f(g(a)))

const double = a => a * 2
const square = a => a * a

module.exports = compose(square, id, double)(computeOne)

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
module.exports = 4;

`;

exports[`32. preval 1`] = `

// @preval
const fs = require('fs')
module.exports = fs.readFileSync(require.resolve('./fixtures/fixture1.md'), 'utf8')

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
module.exports = "# fixture\\n\\nThis is some file thing...\\n";

`;

exports[`33. preval 1`] = `

// @preval
function fib(x) {
  return x <= 1 ? x : fib(x - 1) + fib(x - 2);
}
module.exports = fib(10)

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
module.exports = 55;

`;

exports[`35. preval 1`] = `

// @preval
let smth = {}
module.exports = smth.UNDEFINED;

      ↓ ↓ ↓ ↓ ↓ ↓

// this file was prevaled
module.exports = undefined;

`;
